Add compatibility with tower_http::Body types

samscott89 opened the pull request:
This is a first pass at fixing #19. 

The basic idea was sketched out with @LucioFranco, but I went off on a different path.

What this does is makes `hyper::body::Payload` and `tower_http::Body` fairly interchangeable, by wrapping either one in `LiftBody`. (One direction is already done since `LiftBody` implements `BufStream` which implements `Body`).

I made a few tweaks to the client and (breaking) changes to the server to support this. The result is that the `tower_hyper::client::Connection` can additionally accept any request `B` such that `LiftBody<B>: Payload`.

Services passed to the `tower_hyper::server::Server` need to instead take `LiftBody<Body>` (which is better, since now any blanket request impl taking `Body` works, as in the example), but can return any `B: Body`. So the return types can be more flexible.

samscott89 committed:
Implement `Payload` for lifted `http::Body`s.

Additionally implements `Service<B>` when `LiftBody<B>` is `Payload`.
Enables using `Connection` with anything which implements
`http::Body`, where the return type is `LiftBody<hyper::Body>`,
which implements `BufStream`, and thus implements `http::Body`.

samscott89 committed:
Relax `Server` to take any service producing suitable bodies.

I.e. add an additional generic `B: tower_http::Body`.
Requires implementors to use `LiftBody` for request types,
but more flexibility on responses.

samscott89 committed:
Tweak example to show ability to relax request type.

samscott89 committed:
Remove error println.

LucioFranco commented on src/client/connection.rs:
```diff
@@ -42,3 +43,22 @@ where
     }
 }
+
+impl<B> Service<Request<B>> for Connection<LiftBody<B>>
```
I believe here we just want connection to be generic on `B` only where `B: Body`. We also then want `type Response = Response<B>`

samscott89 committed:
Remove `LiftBody` from API surface as much as possible.

Change Client, Connection to be generic over `HttpBody` instead.

LucioFranco commented on src/client/mod.rs:
```diff
@@ -27,27 +29,34 @@ use tower_service::Service;
-impl<C, B> Client<C, B> {
-    /// Create a new client from a `hyper::Client`
-    pub fn new(inner: hyper::Client<C, B>) -> Self {
+impl<B> Client<HttpConnector, B>
```
this statement shoudl be ->

```rust
impl<C, B> Client<C, B> {
	pub fn new(client: Client<C, B>) -> Self { ... }
}
```

samscott89 replied:
I think it makes sense to have both. `new` being the simplest one which matches `hyper::Client::new()`, and then `with_hyper` or `with_client` to construct one with a particular builder/client.

LucioFranco replied:
Im fine with that 👍 or even, a builder fn that just exposes a special builder?

samscott89 replied:
I've gone with `with_client` for now. The important bit is that the user inputs a `hyper::Client<C, B>` where the `B` part is actually left to be inferred, since we actually require this to be `LiftBody<hyper::Body>`.

LucioFranco commented on src/client/mod.rs:
```diff
@@ -27,27 +29,34 @@ use tower_service::Service;
-    B: Payload + Send + 'static,
-    B::Data: Send,
+    B: HttpBody,
+    LiftBody<B>: hyper::body::Payload,
```
I believe that we can drop this as if B impls HttpBody then we can imply that LiftBody impls Payload I believe.

samscott89 replied:
Currently the conditions on lifting body to payload is:

```rust
    where T: HttpBody + Send + 'static,
          T::Item: Send,
          T::Error: std::error::Error + Send + Sync,
```
Using `LiftBody<B>: Payload` is effectively a shorthand for all of that.

LucioFranco replied:
right we should push those constraints as low as possible

samscott89 replied:
I see your point. I don't think it will be possible to scrub them entirely, but perhaps at the outside-user facing places, it should be the unrolled versions? So that the LiftBody parts aren't exposed in the api?

LucioFranco replied:
yeah client wise that should be possible, I want the client to just pass a `Body` and under the hood we should be able to swap that for a payload without them knowing. And then we can do the same on the other end.

samscott89 replied:
Okay, switched everything to using the trait bounds for clarity on the user-facing structs/traits.

samscott89 commented:
Tests failing because I haven't bothered to fix the retry stuff yet until this all settles.

samscott89 committed:
Simplify trait bounds where possible, add builder for Client.

samscott89 committed:
Allow `RetryPolicy` to use `tower_http::Body` traits instead.

Plus a little bit of tweaking to allow use of `retry::Body`.

samscott89 committed:
Fix doctests.

LucioFranco commented on examples/server.rs:
```diff
@@ -37,17 +36,20 @@ fn main() {
-        let body = LiftBody::new(Body::from("Hello World!"));
+    fn call(&mut self, _req: Request<B>) -> Self::Future {
+        // let body = LiftBody::new(Body::from("Hello World!"));
+        let body = b"Hello World!".to_vec();
```
minor thing might be nice to show this as `req.into_body()` then pass that back.

LucioFranco commented on src/body.rs:
```diff
@@ -24,11 +24,38 @@ impl<T: Payload> LiftBody<T> {
+impl<T> Payload for LiftBody<T>
+    where T: HttpBody + Send + 'static,
+          T::Item: Send,
+          T::Error: Into<Box<std::error::Error + Send + Sync>>,
```
minor nit: can we move the `Box<...>` to a `pub(crate) type Error = Box<...>`, this is just a pattern we have been using throughout tower.

LucioFranco commented on src/client/connect.rs:
```diff
@@ -82,7 +83,8 @@ where
     C: MakeConnection<A> + 'static,
-    B: Payload + 'static,
+    B: HttpBody,
+    LiftBody<B>: hyper::body::Payload,
```
should be able to drop this

LucioFranco commented on src/client/connect.rs:
```diff
@@ -110,7 +112,8 @@ where
     C: MakeConnection<A>,
-    B: Payload,
+    B: HttpBody,
+    LiftBody<B>: hyper::body::Payload,
```
and this

samscott89 replied:
You really wont let me have my shorthand bounds, will you D:

LucioFranco commented on src/client/connection.rs:
```diff
@@ -12,33 +15,61 @@ use tower_service::Service;
+}
+
+/// Lift a hyper ResponseFuture to one which returns LiftBody
+pub struct LiftResponseFuture<F>(pub(crate) F);
```
we could prob just name this `ResponseFuture`

LucioFranco commented on src/client/mod.rs:
```diff
@@ -42,12 +83,13 @@ where
-    B::Data: Send,
+    B: HttpBody + Send + 'static,
+    B::Item: Send,
+    B::Error: Into<Box<std::error::Error + Send + Sync>>,
```
same here for the error type

LucioFranco reviewed (approved):
This looks fantastic!! Really really happy with how this came out, great work! I ahve just a few minor comments left but otherwise LGTM 👍

samscott89 committed:
Minor tweaks.

Use `crate::Error` for consistent error typing.
Tweak trait bounds.
Rename LiftResponseFuture.

LucioFranco commented:
This is great! Thanks @samscott89 great work 👍
